lib/Jsrt/Jsrt.cpp

```diff
@@ -4971,25 +4971,21 @@ CHAKRA_API JsSerialize(
 
     *bufferVal = nullptr;
 
-    bool isExternalArray = Js::ExternalArrayBuffer::Is(scriptVal),
-         isString = false;
-    bool isUtf8   = !(parseAttributes & JsParseScriptAttributeArrayBufferIsUtf16Encoded);
-    if (!isExternalArray)
+    const bool isExternalArray = Js::ExternalArrayBuffer::Is(scriptVal);
+    const bool isString = !isExternalArray && Js::JavascriptString::Is(scriptVal);
+    if (!isExternalArray && !isString)
     {
-        isString = Js::JavascriptString::Is(scriptVal);
-        if (!isString)
-        {
-            return JsErrorInvalidArgument;
-        }
+        return JsErrorInvalidArgument;
     }
+    const bool isUtf8 = !isString && !(parseAttributes & JsParseScriptAttributeArrayBufferIsUtf16Encoded);
 
     LoadScriptFlag scriptFlag;
     const byte* script = isExternalArray ?
         ((Js::ExternalArrayBuffer*)(scriptVal))->GetBuffer() :
         (const byte*)((Js::JavascriptString*)(scriptVal))->GetSz();
     const size_t cb = isExternalArray ?
         ((Js::ExternalArrayBuffer*)(scriptVal))->GetByteLength() :
-        ((Js::JavascriptString*)(scriptVal))->GetLength();
+        ((Js::JavascriptString*)(scriptVal))->GetAllocatedByteCount();
```

Comment:
I didn't realize we had this function. I don't think that this is always correct though; looks like a substring can report 0 for this... Might be safer to just use `->GetLength() * 2`? Although the call to `GetSz` just before will flatten strings and might mean that this ends up behaving as expected.

Reply:
What would it mean to have scriptVal as SubString here? How could I test it? From the general design point of view, if GetAllocatedByteCount doesn't work for subclasses they must override it (which SubString does, but I don't understand semantics: why is it correct to return 0 if still have original reference?). I'd rather not do math manually here because it will become wrong when we enable narrow strings (and hard to spot that it needs to be fixed).

---
In reply to: [177916631](https://github.com/Microsoft/ChakraCore/pull/4896#discussion_r177916631) [](ancestors = 177916631)

Reply:
My hypothetical scenario would be if you retrieved a string from script instead of explicitly creating it (e.g. you run a function like `function () { return "1234".substring(2);}`) and then pass that value into `JsSerialize`. I don't endorse this, nor do I expect it, but right now it's at least possible.

Considering that `JavascriptString::GetAllocatedByteCount` also returns 0 in the case when it isn't finalized (essentially if it isn't yet flattened into a single buffer; we have a number of tree-like structures for representing intermediate string forms) it may be worth addressing.... but also considering that `JavascriptString::GetSz()` flattens and finalizes strings, as long as we `GetSz` before checking the allocated bytes it may be ok.

Reply:
Added JavascriptString::GetSizeInBytes

---
In reply to: [177919117](https://github.com/Microsoft/ChakraCore/pull/4896#discussion_r177919117) [](ancestors = 177919117)